src/iterators.rs

```diff
@@ -550,3 +556,538 @@ impl PyGCProtocol for WeightedEdgeList {
+                            other_raw.downcast::<PySequence>()?;
+                        if value.len() as isize != other_value.len()? {
+                            return Ok(false);
+                        }
+                        for (i, item) in value.iter().enumerate() {
+                            let other_item_raw =
+                                other_value.get_item(i as isize)?;
+                            let other_item_value: usize =
+                                other_item_raw.extract()?;
+                            if other_item_value != *item {
+                                return Ok(false);
+                            }
+                        }
+                    }
+                    Err(ref err)
+                        if Python::with_gil(|py| {
+                            err.is_instance::<PyKeyError>(py)
+                        }) =>
+                    {
+                        return Ok(false);
+                    }
+                    Err(err) => return Err(err),
+                }
+            }
+            Ok(true)
+        };
+        match op {
+            pyo3::basic::CompareOp::Eq => compare(other),
+            pyo3::basic::CompareOp::Ne => match compare(other) {
+                Ok(res) => Ok(!res),
+                Err(err) => Err(err),
+            },
+            _ => Err(PyNotImplementedError::new_err(
+                "Comparison not implemented",
+            )),
+        }
+    }
+
+    fn __str__(&self) -> PyResult<String> {
+        let mut str_vec: Vec<String> = Vec::with_capacity(self.paths.len());
+        for path in &self.paths {
+            str_vec.push(format!(
+                "{}: {}",
+                path.0,
+                format!(
+                    "[{}]",
+                    path.1
+                        .iter()
+                        .map(|n| format!("{}", n))
+                        .collect::<Vec<String>>()
+                        .join(", ")
+                ),
+            ));
+        }
+        Ok(format!("PathMapping{{{}}}", str_vec.join(", ")))
+    }
+
+    fn __hash__(&self) -> PyResult<u64> {
+        let mut hasher = DefaultHasher::new();
+        for index in &self.paths {
+            hasher.write_usize(*index.0);
+            for node in index.1 {
+                hasher.write_usize(*node);
+            }
+        }
+        Ok(hasher.finish())
+    }
+}
+
+#[pyproto]
+impl PyMappingProtocol for PathMapping {
+    /// Return the number of nodes in the graph
+    fn __len__(&self) -> PyResult<usize> {
+        Ok(self.paths.len())
+    }
+    fn __getitem__(&'p self, idx: usize) -> PyResult<NodeIndices> {
+        match self.paths.get(&idx) {
+            Some(data) => Ok(NodeIndices {
+                nodes: data.clone(),
+            }),
+            None => Err(PyIndexError::new_err("No node found for index")),
+        }
+    }
+}
+
+#[pyproto]
+impl PySequenceProtocol for PathMapping {
+    fn __len__(&self) -> PyResult<usize> {
```

Comment:
it's interesting that this has to be implemented multiple times depending on the protocol! Am I reading that right? (this line and 727) This means these could theoretically be implemented differently depending on the context?

Reply:
Yeah, it's weird. But looking at the Python C API docs these are just the rust methods wrapping 2 different functions in the C protocols api:

https://docs.python.org/3/c-api/sequence.html#c.PySequence_Length
https://docs.python.org/3/c-api/mapping.html#c.PyMapping_Length

I'm not sure what circumstances one would be called over the other tbh, I was just implementing the required methods for the protocol `impl`.